letsencrypt/models/letsencrypt.py

```diff
@@ -2,167 +2,223 @@
-                ])
-            self.call_cmdline(cmdline)
-        return csr
+                    "or local domains!"))
 
     @api.model
-    def cron(self):
+    def _cron(self):
         domain = urlparse.urlparse(
             self.env['ir.config_parameter'].get_param(
                 'web.base.url', 'localhost')).netloc
-        self.validate_domain(domain)
-        account_key = self.generate_account_key()
-        csr = self.generate_csr(domain)
-        acme_challenge = get_challenge_dir()
-        if not os.path.isdir(acme_challenge):
-            os.makedirs(acme_challenge)
-        if self.env.context.get('letsencrypt_dry_run'):
-            crt_text = 'I\'m a test text'
-        else:  # pragma: no cover
-            from acme_tiny import get_crt, DEFAULT_CA
-            crt_text = get_crt(
-                account_key, csr, acme_challenge, log=_logger, CA=DEFAULT_CA)
-        with open(os.path.join(get_data_dir(), '%s.crt' % domain), 'w')\
-                as crt:
-            crt.write(crt_text)
-            chain_cert = urllib2.urlopen(
+        self._validate_domain(domain)
+        # Generate account key
+        account_key_file = self._generate_key('account.key')
+        account_key = jose.JWKRSA.load(open(account_key_file).read())
+        # Generate domain key
+        domain_key_file = self._generate_key(domain)
+        client = self._create_client(account_key)
+        new_reg = NewRegistration(
+            key=account_key.public_key(),
+            terms_of_service_agreed=True)
+        client.new_account(new_reg)
+        csr = self._make_csr(account_key, domain_key_file, domain)
+        authzr = client.new_order(csr)
+        # write the file with the challenge
+        challenge_dir = _get_challenge_dir()
+        if not isdir(challenge_dir):
+            makedirs(challenge_dir)
+        auth_responded = False
+        for authorizations in authzr.authorizations:
+            for challenge in sorted(
+                    authorizations.body.challenges, key=lambda x: x != 'http'):
+                if challenge.chall.typ == TYPE_CHALLENGE_HTTP:
+                    self._respond_challenge_http(challenge, account_key)
+                    client.answer_challenge(
+                        challenge,
+                        challenges.HTTP01Response())
+                    auth_responded = True
+                    break
+                elif challenge.chall.typ == TYPE_CHALLENGE_DNS:
+                    self._respond_challenge_dns(challenge, domain)
+                    client.answer_challenge(
+                        challenge, challenges.DNSResponse())
+                    auth_responded = True
+                    break
+        if not auth_responded:
+            raise exceptions.ValidationError(
+                _('Could not respond to letsencrypt challenges.'))
+        # let them know we are done and they should check
+        order_resource = client.poll_and_finalize(authzr)
```

Comment:
it might also be worthwhile to check in the dns part above if the record is created correctly, wait for a few seconds, check again until some timeout is reached

Reply:
There already exist default for this https://github.com/certbot/certbot/blob/0.27.x/acme/acme/client.py#L659 aren't these enough?

Reply:
run it on our test-instance, there it came back with an error immediately for me at least. You might have to pass a much longer deadline, but try first